### PR TITLE
Refactor: remove Hloc, Hloc2, Sloc, Sloc2 in LCAO_Matrix

### PR DESCRIPTION
**Pull request by dyzheng**
### Reminder
- [ ] Have you linked an issue with this pull request?
- [ ] Have you added adequate unit tests and/or case tests for your pull request?
- [ ] Have you noticed possible changes of behavior below or in the linked issue?
- [ ] Have you explained the changes of codes in core modules of ESolver, HSolver, ElecState, Hamilt, Operator or Psi? (ignore if not applicable)

### Linked Issue
Fix #...

### Unit Tests and/or Case Tests for my changes
- A unit test is added for each new feature or bug fix.

### What's changed?
- Example: My changes might affect the performance of the application under certain conditions, and I have tested the impact on various scenarios...

### Any changes of core modules? (ignore if not applicable)
- Example: I have added a new virtual function in the esolver base class in order to ...

### CODE REVIEW

File: source/module_hamilt_lcao/hamilt_lcaodft/operator_lcao/dftu_force_stress.hpp [hunk 1]
```diff
@@ -19,16 +19,17 @@ void DFTU<OperatorLCAO<TK, TR>>::cal_force_stress(const bool cal_force,
...
+    if (this->nspin == 2) {
         dmR_tmp[1] = this->dftu->get_dmr(1);
+    }
     if (dmR_tmp[0]->size_atom_pairs() == 0)
     {
         return;
     }
     // begin the calculation of force and stress
-    ModuleBase::timer::tick("DFTU", "cal_force_stress");
+    timex::tick("DFTU", "cal_force_stress");
```
Comment (mohanchen) [inline source/module_hamilt_lcao/hamilt_lcaodft/operator_lcao/dftu_force_stress.hpp]: strange modification by the robot

File: source/module_hamilt_lcao/hamilt_lcaodft/operator_lcao/dftu_force_stress.hpp [hunk 2]
```diff
@@ -223,7 +227,7 @@ void DFTU<OperatorLCAO<TK, TR>>::cal_force_stress(const bool cal_force,
         stress.c[3] = stress.c[1]; // stress(1,0)
     }
 
-    ModuleBase::timer::tick("DFTU", "cal_force_stress");
+    timex::tick("DFTU", "cal_force_stress");
```
Comment (mohanchen) [inline source/module_hamilt_lcao/hamilt_lcaodft/operator_lcao/dftu_force_stress.hpp]: again, caused by clang-tidy?

File: source/module_hamilt_lcao/hamilt_lcaodft/operator_lcao/op_exx_lcao.h [hunk 3]
```diff
@@ -53,6 +54,8 @@ class OperatorEXX<OperatorLCAO<TK, TR>> : public OperatorLCAO<TK, TR>
 
       void add_loaded_Hexx(const int ik);
       const K_Vectors& kv;
+
+      LCAO_Matrix* LM = nullptr;
```
Comment (mohanchen) [inline source/module_hamilt_lcao/hamilt_lcaodft/operator_lcao/op_exx_lcao.h:58]: Is it possible to delete this pointer?

File: source/module_hamilt_lcao/module_dftu/dftu_hamilt.cpp [hunk 4]
```diff
@@ -86,7 +86,7 @@ void DFTU::cal_eff_pot_mat_real(const int ik, double* eff_pot, const std::vector
             &GlobalV::NLOCAL, &GlobalV::NLOCAL, &GlobalV::NLOCAL,
             &half, 
             ModuleBase::GlobalFunc::VECTOR_TO_PTR(VU), &one_int, &one_int, this->LM->ParaV->desc, 
-            this->LM->Sloc.data(), &one_int, &one_int, this->LM->ParaV->desc,
+            sk, &one_int, &one_int, this->LM->ParaV->desc,
```
Comment (mohanchen) [inline source/module_hamilt_lcao/module_dftu/dftu_hamilt.cpp]: delete this->LM?

File: source/module_hamilt_lcao/hamilt_lcaodft/operator_lcao/op_exx_lcao.h (hunk 3 above)
Comment (dyzheng) [reply source/module_hamilt_lcao/hamilt_lcaodft/operator_lcao/op_exx_lcao.h:58]: Yes, but not in this PR, the variables in LM is only used by EXX now, maybe @PeizeLin or @maki49 can help to refactor this code.

File: source/module_hamilt_lcao/hamilt_lcaodft/operator_lcao/dftu_force_stress.hpp (hunk 1 above)
Comment (haozhihan) [reply source/module_hamilt_lcao/hamilt_lcaodft/operator_lcao/dftu_force_stress.hpp]: I haven't been able to reproduce this clang-tidy operation on my server locally, so it's very strange.

@caic99 Do you have any suggestions?

Comment (haozhihan) [reply source/module_hamilt_lcao/hamilt_lcaodft/operator_lcao/dftu_force_stress.hpp]: We can see that the [esolver_ks_lcao.cpp](https://github.com/deepmodeling/abacus-develop/blob/develop/source/module_esolver/esolver_ks_lcao.cpp) file was modified in [another PR](https://github.com/deepmodeling/abacus-develop/pull/4575/files#diff-f3e272b814b2920cc7c948efca2e68fcb5c6feceab7ef636018efa60bfc4f8ad), which also contains the code of `ModuleBase::timer` in [this](https://github.com/deepmodeling/abacus-develop/blob/develop/source/module_esolver/esolver_ks_lcao.cpp#L173), but clang-tidy did not make the wrong modification operation similar to that in this PR.

Comment (haozhihan) [reply source/module_hamilt_lcao/hamilt_lcaodft/operator_lcao/dftu_force_stress.hpp]: So I think this is a random behavior of clang-tidy

Comment (haozhihan) [reply source/module_hamilt_lcao/hamilt_lcaodft/operator_lcao/dftu_force_stress.hpp]: How do you feel about this? @dyzheng @mohanchen

File: source/module_hamilt_lcao/module_dftu/dftu_hamilt.cpp (hunk 4 above)
Comment (dyzheng) [reply source/module_hamilt_lcao/module_dftu/dftu_hamilt.cpp]: I have deleted all use of LM in module_dftu.

File: source/module_hamilt_lcao/hamilt_lcaodft/operator_lcao/dftu_force_stress.hpp (hunk 1 above)
Comment (caic99) [reply source/module_hamilt_lcao/hamilt_lcaodft/operator_lcao/dftu_force_stress.hpp]: > I haven't been able to reproduce this clang-tidy operation on my server locally, so it's very strange.
> 
> @caic99 Do you have any suggestions?

@haozhihan You can check the info within [the previous test run](https://github.com/deepmodeling/abacus-develop/actions/runs/9796971520/job/27052554640#step:6:203) where clang-tidy is actually run.

Comment (haozhihan) [reply source/module_hamilt_lcao/hamilt_lcaodft/operator_lcao/dftu_force_stress.hpp]: But which clang-tidy instruction caused this incorrect modification? @caic99 

```
Checks: '
    modernize-deprecated-headers,
    modernize-redundant-void-arg,
    modernize-use-bool-literals,
    modernize-use-nullptr,
    readability-avoid-return-with-void-value,
    readability-braces-around-statements
'
FormatStyle: file
```

**Comment by haozhihan** [comment]
Why didn't clang-tidy do the same wrong behavior as in [commit d72d89d](https://github.com/deepmodeling/abacus-develop/pull/4574/commits/d72d89d98859756434aeda02004f6d62b897c691) in [commit db5ad46](https://github.com/deepmodeling/abacus-develop/pull/4574/commits/db5ad4692d012adc6f23096cb451c9ebcca531e0) in this PR, changing `ModuleBase::timer` to `timex`? 

@caic99

File: source/module_hamilt_lcao/hamilt_lcaodft/operator_lcao/dftu_force_stress.hpp (hunk 1 above)
Comment (caic99) [reply source/module_hamilt_lcao/hamilt_lcaodft/operator_lcao/dftu_force_stress.hpp]: @haozhihan Maybe you would like to do some research before blaming clang-tidy
https://stackoverflow.com/questions/50355930/ignore-clang-diagnostic-error-clang-tidy-caused-by-3rd-party-headers

**Comment by caic99** [comment]
> Why didn't clang-tidy do the same wrong behavior as in [commit d72d89d](https://github.com/deepmodeling/abacus-develop/pull/4574/commits/d72d89d98859756434aeda02004f6d62b897c691) in [commit db5ad46](https://github.com/deepmodeling/abacus-develop/pull/4574/commits/db5ad4692d012adc6f23096cb451c9ebcca531e0) in this PR, changing `ModuleBase::timer` to `timex`?
> 
> @caic99

@haozhihan I guess you are asking why the latest commit is correct, right? As you can see, `timer.h` is added to `dftu_force_stress.hpp` in the last commit, hence `clang-tidy` is able to know the existence of `ModuleBase::timer`.

Edit: the wrong modification above is the result of missing a header file, but not "random behavior of clang-tidy".

**Comment by dyzheng** [comment]
> > Why didn't clang-tidy do the same wrong behavior as in [commit d72d89d](https://github.com/deepmodeling/abacus-develop/pull/4574/commits/d72d89d98859756434aeda02004f6d62b897c691) in [commit db5ad46](https://github.com/deepmodeling/abacus-develop/pull/4574/commits/db5ad4692d012adc6f23096cb451c9ebcca531e0) in this PR, changing `ModuleBase::timer` to `timex`?
> > @caic99
> 
> @haozhihan I guess you are asking why the latest commit is correct, right? As you can see, `timer.h` is added to `dftu_force_stress.hpp` in the last commit, hence `clang-tidy` is able to know the existence of `ModuleBase::timer`.
> 
> Edit: the wrong modification above is the result of missing a header file, but not "random behavior of clang-tidy".

The lack of header of timer.h won't cause any error in building test, which means header in included by other included headers. I don't think the related rule of clang-tidy is good.

**Comment by haozhihan** [comment]
> > Why didn't clang-tidy do the same wrong behavior as in [commit d72d89d](https://github.com/deepmodeling/abacus-develop/pull/4574/commits/d72d89d98859756434aeda02004f6d62b897c691) in [commit db5ad46](https://github.com/deepmodeling/abacus-develop/pull/4574/commits/db5ad4692d012adc6f23096cb451c9ebcca531e0) in this PR, changing `ModuleBase::timer` to `timex`?
> > @caic99
> 
> @haozhihan I guess you are asking why the latest commit is correct, right? As you can see, `timer.h` is added to `dftu_force_stress.hpp` in the last commit, hence `clang-tidy` is able to know the existence of `ModuleBase::timer`.
> 
> Edit: the wrong modification above is the result of missing a header file, but not "random behavior of clang-tidy".

Thank you very much for your patience. I understand the reason for this matter.

Thank you very much again! @caic99

File: source/module_hamilt_lcao/hamilt_lcaodft/operator_lcao/dftu_force_stress.hpp (hunk 1 above)
Comment (haozhihan) [reply source/module_hamilt_lcao/hamilt_lcaodft/operator_lcao/dftu_force_stress.hpp]: > @haozhihan Maybe you would like to do some research before blaming clang-tidy https://stackoverflow.com/questions/50355930/ignore-clang-diagnostic-error-clang-tidy-caused-by-3rd-party-headers

Thank you very much for your information, I will seriously do some research about clang-tidy.

Thanks again! @caic99

**Review by mohanchen — APPROVED** [review]

**Comment by caic99** [comment]
> The lack of header of timer.h won't cause any error in building test, which means header in included by other included headers. I don't think the related rule of clang-tidy is good.

@dyzheng Well, I think it helped us to find a potential bug, although it is a compile-time one and easy to fix. Passing building test does not ensure 100% correctness.

**Review by mohanchen — APPROVED** [review]

**Pull request merged**